e2e/internal/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,341 @@
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/edgelesssys/constellation/v2/e2e/internal/kubectl"
+	"github.com/edgelesssys/constellation/v2/internal/attestation/measurements"
+	"github.com/edgelesssys/constellation/v2/internal/cloud/cloudprovider"
+	"github.com/edgelesssys/constellation/v2/internal/config"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/versionsapi"
+	"github.com/edgelesssys/constellation/v2/internal/versionsapi/fetcher"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	constellationCLIVersion = "2.5.1"
+	wantWorkerNodeCount     = 2
+	wantControlNodeCount    = 3
+	wantNodeCount           = wantControlNodeCount + wantWorkerNodeCount
+)
+
+// E2E is creating cluster from root of Constellation repository, we need to
+// execute the upgrade test in that Constellation workspace.
+func TestMain(m *testing.M) {
+	if err := os.Chdir("../../.."); err != nil {
+		os.Exit(1)
+	}
+
+	os.Exit(m.Run())
+}
+
+func mustDecodeHex(s string) [32]byte {
+	val, _ := hex.DecodeString(s)
+	return *(*[32]byte)(val)
+}
+
+func TestUpgrade(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	k, err := kubectl.New()
+	require.NoError(err)
+	assert.NotNil(k)
+
+	testNodesEventuallyAvailable(t, k, wantControlNodeCount, wantWorkerNodeCount)
+	testPodsEventuallyReady(t, k, "kube-system")
+
+	testCLIHasVersion(t, constellationCLIVersion)
+	wantImage := writeUpgradeConfig(t)
+	cmd := exec.CommandContext(context.Background(), "constellation", "upgrade", "execute")
+	msg, err := cmd.CombinedOutput()
+	require.NoErrorf(err, "%s", string(msg))
+
+	testNodesEventuallyHaveImage(t, k, wantImage)
+}
+
+func writeUpgradeConfig(t *testing.T) string {
+	fileHandler := file.NewHandler(afero.NewOsFs())
+	cfg, err := config.New(fileHandler, constants.ConfigFilename)
+	require.NoError(t, err)
+
+	info, err := fetchLatestNightlyUpgradeInfo(cfg.GetProvider())
+	require.NoError(t, err)
+
+	cfg.Upgrade.Image = info.shortPath
+	cfg.Upgrade.Measurements = make(measurements.M)
+	for key, value := range info.measurements {
+		cfg.Upgrade.Measurements[key] = measurements.Measurement{
+			Expected: mustDecodeHex(value),
+			WarnOnly: false,
+		}
+	}
+	err = fileHandler.WriteYAML(constants.ConfigFilename, cfg, file.OptOverwrite)
+	require.NoError(t, err)
+
+	return info.wantImage
+}
+
+func testNodesEventuallyHaveImage(t *testing.T, k *kubernetes.Clientset, wantImage string) {
+	require.Eventually(t, func() bool {
+		nodes, err := k.CoreV1().Nodes().List(context.Background(), metaV1.ListOptions{})
+		if err != nil {
+			fmt.Println(err)
+			return false
+		}
+
+		allUpdated := true
+		fmt.Printf("Node status (%v):\n", time.Now())
```

Comment:
```suggestion
		t.Log(fmt.Sprintf("Node status (%v):", time.Now()))
```
Might be better to use t.Log ?